File: backend/llm.py
import asyncio
import os

import httpx

OPENROUTER_URL = "https://openrouter.ai/api/v1/chat/completions"


def _headers() -> dict:
    api_key = os.environ["OPENROUTER_API_KEY"]
    return {
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json",
        # OpenRouter uses these two headers for attribution / free-tier routing.
        "HTTP-Referer": os.environ.get("FRONTEND_ORIGIN", "http://localhost:3000"),
        "X-Title": "Multi-Agent Research Assistant",
    }


async def call_llm(
    system_prompt: str,
    user_prompt: str,
    temperature: float = 0.4,
    max_retries: int = 3,
) -> str:
    """Call an OpenRouter chat model, retrying transient failures with backoff.

    Free-tier models return 429 when overloaded or rate-limited; a short
    exponential backoff (honoring Retry-After when present) usually clears it.
    """
    model = os.environ.get("OPENROUTER_MODEL", "openrouter/free")
    body = {
        "model": model,
        "messages": [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_prompt},
        ],
        "temperature": temperature,
    }

    last_exc: Exception | None = None
    for attempt in range(max_retries):
        try:
            async with httpx.AsyncClient(timeout=90) as client:
                resp = await client.post(OPENROUTER_URL, headers=_headers(), json=body)
                resp.raise_for_status()
                data = resp.json()
                return data["choices"][0]["message"]["content"]
        except (httpx.HTTPStatusError, httpx.TransportError) as e:
            last_exc = e
            if attempt == max_retries - 1:
                break
            delay = 2.0 * (2**attempt)  # 2s, 4s, then give up
            response = getattr(e, "response", None)
            if response is not None and response.status_code == 429:
                retry_after = response.headers.get("Retry-After")
                if retry_after and retry_after.isdigit():
                    delay = max(delay, min(float(retry_after), 30.0))
            await asyncio.sleep(delay)

    raise last_exc  # type: ignore[misc] - always set after the loop


def clean_json_block(raw: str) -> str:
    """Strips markdown code fences if the model wraps its JSON in ```json ... ```"""
    text = raw.strip()
    if text.startswith("```"):
        parts = text.split("```")
        text = parts[1]
        if text.startswith("json"):
            text = text[4:]
    return text.strip()

File: supabase/functions/_shared/llm.ts
/**
 * LLM access for the agents.
 *
 * Primary provider: Hugging Face Inference (OpenAI-compatible router) using the
 * account's HF token — free, generous quota, no credit card.
 * Fallback provider: OpenRouter (same API shape) using the project's existing key.
 */

export interface LLMMessage {
  role: "system" | "user";
  content: string;
}

async function postChat(
  url: string,
  headers: Record<string, string>,
  body: unknown,
): Promise<string> {
  let lastErr: unknown;
  for (let attempt = 0; attempt < 3; attempt++) {
    try {
      const resp = await fetch(url, { method: "POST", headers, body: JSON.stringify(body) });
      if (resp.status === 429 || resp.status >= 500) {
        const retryAfter = resp.headers.get("retry-after");
        const base = 2 * 2 ** attempt;
        const delay = retryAfter && !Number.isNaN(Number(retryAfter))
          ? Math.min(Number(retryAfter), 30)
          : base;
        await new Promise((r) => setTimeout(r, delay * 1000));
        continue;
      }
      if (!resp.ok) {
        const text = (await resp.text()).slice(0, 300);
        throw new Error(`LLM ${resp.status}: ${text}`);
      }
      const data = await resp.json();
      return data.choices?.[0]?.message?.content ?? "";
    } catch (e) {
      lastErr = e;
      if (attempt < 2) await new Promise((r) => setTimeout(r, 2 * 2 ** attempt * 1000));
    }
  }
  throw lastErr instanceof Error ? lastErr : new Error(String(lastErr));
}

export async function callLLM(
  system: string,
  user: string,
  opts: { maxTokens?: number; temperature?: number } = {},
): Promise<string> {
  const messages: LLMMessage[] = [
    { role: "system", content: system },
    { role: "user", content: user },
  ];
  const temperature = opts.temperature ?? 0.4;
  const max_tokens = opts.maxTokens ?? 1200;

  const hfToken = Deno.env.get("HF_TOKEN");
  if (hfToken) {
    const model = Deno.env.get("HF_MODEL") || "Qwen/Qwen2.5-7B-Instruct";
    try {
      return await postChat(
        "https://router.huggingface.co/v1/chat/completions",
        { Authorization: `Bearer ${hfToken}`, "Content-Type": "application/json" },
        { model, messages, temperature, max_tokens },
      );
    } catch (e) {
      console.error("HF inference failed, falling back to OpenRouter:", String(e));
    }
  }

  const orKey = Deno.env.get("OPENROUTER_API_KEY");
  if (orKey) {
    const model = Deno.env.get("OPENROUTER_MODEL") || "openrouter/free";
    return await postChat(
      "https://openrouter.ai/api/v1/chat/completions",
      {
        Authorization: `Bearer ${orKey}`,
        "Content-Type": "application/json",
        "HTTP-Referer": "https://huggingface.co",
        "X-Title": "Multi-Agent Research Assistant",
      },
      { model, messages, temperature, max_tokens },
    );
  }

  throw new Error("No LLM provider configured (set HF_TOKEN or OPENROUTER_API_KEY)");
}
